### PR TITLE
HDFS-15079. RBF: Namenode needs to use the actual client Id and callId when going through RBF proxy.

### PR DESCRIPTION
**Pull request by ZanderXu**
### Description of PR
Jira: [HDFS-15079](https://issues.apache.org/jira/browse/HDFS-15079)

Similarly with [HDFS-13248](https://issues.apache.org/jira/browse/HDFS-13248),  RBF adds the actual client Id and client call Id in CallerContext and carries them to NameNode. Then nameNode try to obtain the actual client Id and client call Id to ensure CacheEntry mechanism.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m 12s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 3 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 35s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  30m 24s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  27m  3s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  compile  |  21m 55s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   4m 32s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 17s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   4m 16s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 40s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   8m 44s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  24m 55s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 10s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  24m 23s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javac  |  24m 23s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 50s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  21m 50s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m 25s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/1/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 1 new + 245 unchanged - 6 fixed = 246 total (was 251)  |
| +1 :green_heart: |  mvnsite  |   5m 13s |  |  the patch passed  |
| -1 :x: |  javadoc  |   1m 23s | [/results-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkPrivateBuild-11.0.15+10-Ubuntu-0ubuntu0.20.04.1.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/1/artifact/out/results-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkPrivateBuild-11.0.15+10-Ubuntu-0ubuntu0.20.04.1.txt) |  hadoop-common-project_hadoop-common-jdkPrivateBuild-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 generated 4 new + 0 unchanged - 0 fixed = 4 total (was 0)  |
| +1 :green_heart: |  javadoc  |   4m 35s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   9m 13s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m  2s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 20s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  | 381m 42s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |  35m 19s |  |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 37s |  |  The patch does not generate ASF License warnings.  |
|  |   | 690m 55s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 47101b3a9fc4 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 0813ade6d6810bc69833c55e6369c2084df87dcb |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/1/testReport/ |
| Max. process+thread count | 2792 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  7s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 3 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 37s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  28m 27s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  29m 50s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  compile  |  23m 24s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   4m 36s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 21s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   4m 16s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 36s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   8m 48s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m  3s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 11s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  24m 36s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javac  |  24m 36s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  22m  0s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  22m  0s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   4m 23s |  |  root: The patch generated 0 new + 245 unchanged - 6 fixed = 245 total (was 251)  |
| +1 :green_heart: |  mvnsite  |   5m 15s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   4m  6s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 36s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   9m 18s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m  2s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 16s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  | 397m  8s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |  36m 32s |  |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 38s |  |  The patch does not generate ASF License warnings.  |
|  |   | 710m 22s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux c397390b14f9 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 177049942039c15fb67a7b3e2fafb7e4fcb063f6 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/2/testReport/ |
| Max. process+thread count | 2324 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/retry/RetryInvocationHandler.java [hunk 1]
```diff
@@ -46,6 +46,15 @@
   public static final Logger LOG = LoggerFactory.getLogger(
       RetryInvocationHandler.class);
 
+  @VisibleForTesting
+  public static final ThreadLocal<Boolean> SET_CALL_ID_FOR_TEST =
+      new ThreadLocal<Boolean>() {
```
Comment (goiri) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/retry/RetryInvocationHandler.java]: There's no lambda or shorter version for this?

File: hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterRetryCache.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            new String[]{"fake_group"});
+    FileSystem joeFS = joe.doAs(
+        (PrivilegedExceptionAction<FileSystem>) () ->
+            FileSystem.newInstance(routerFS.getUri(), routerFS.getConf()));
+
+    Path renameSrc = new Path(testDir, "renameSrc");
+    Path renameDst = new Path(testDir, "renameDst");
+    joeFS.mkdirs(renameSrc);
+
+    Assert.assertEquals(HAServiceProtocol.HAServiceState.ACTIVE,
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterRetryCache.java]: You can statically import these asserts and just do `assertEqual()`

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java [hunk 3]
```diff
@@ -495,6 +498,72 @@ public static NameNodeMetrics getNameNodeMetrics() {
     return metrics;
   }
 
+  private static String clientInfoFromContext(
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:505]: One line and javadoc.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java [hunk 4]
```diff
@@ -495,6 +498,72 @@ public static NameNodeMetrics getNameNodeMetrics() {
...
+        CallerContext context = CallerContext.getCurrent();
+        if (context != null && context.isContextValid()) {
+          return context.getContext();
+        }
+      }
+    }
+    return null;
+  }
+
+  private static String parseSpecialValue(String content, String key) {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java]: Can we have unit tests for this?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java [hunk 5]
```diff
@@ -495,6 +498,72 @@ public static NameNodeMetrics getNameNodeMetrics() {
...
+    }
+
+    String clientMachine = Server.getRemoteAddress();
+    if (clientMachine == null) { //not a RPC client
+      clientMachine = "";
+    }
+    return clientMachine;
+  }
+
+  public static Pair<byte[], Integer> getClientIdAndCallId(
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:567]: javadoc

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java [hunk 6]
```diff
@@ -495,6 +498,72 @@ public static NameNodeMetrics getNameNodeMetrics() {
...
+    if (posn != -1) {
+      posn += key.length();
+      int end = content.indexOf(",", posn);
+      return end == -1 ? content.substring(posn)
+          : content.substring(posn, end);
+    }
+    return null;
+  }
+
+  public static String getClientMachine(final String[] ipProxyUsers) {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:544]: javadoc

**Comment by ZanderXu** [comment]
@goiri Thanks for your review and I have updated the patch, please help me review it.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  4s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 3 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 58s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  28m 21s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  25m  5s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  compile  |  21m 55s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   4m 34s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 13s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   4m 16s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 33s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   8m 52s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  24m 46s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 32s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 20s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  24m 57s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javac  |  24m 57s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 59s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  21m 59s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   4m 24s |  |  root: The patch generated 0 new + 245 unchanged - 6 fixed = 245 total (was 251)  |
| +1 :green_heart: |  mvnsite  |   5m 14s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   4m  9s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 36s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   9m  8s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m 36s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 13s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  | 379m 26s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |  34m 32s |  |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 40s |  |  The patch does not generate ASF License warnings.  |
|  |   | 684m 39s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 1b0a6da5dc5f 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / a29bedfeb1fb2f3fd22a525632e03ebd1d357c0d |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/3/testReport/ |
| Max. process+thread count | 2329 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ZanderXu** [comment]
@goiri @ferhui @Hexiaoqiao @ayushtkn can you help me to review this patch?

If  you have any other ideas, please let me know. This is a very helpful issue for end users, so I want to push it forward.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/RetryCache.java [hunk 7]
```diff
@@ -140,7 +140,7 @@ public long getExpirationTime() {
     
     @Override
     public String toString() {
-      return (new UUID(this.clientIdMsb, this.clientIdLsb)).toString() + ":"
+      return (new UUID(this.clientIdMsb, this.clientIdLsb)) + ":"
```
Comment (goiri) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/RetryCache.java]: Maybe is cleaner to have this as String.format()

**Review by goiri — APPROVED** [review]

**Comment by goiri** [comment]
It would be nice to get feedback from others.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  1s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 3 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 28s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  28m 23s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  25m 12s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  compile  |  22m 29s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   5m 14s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   6m  7s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   5m 27s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   5m 16s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   9m  0s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  24m 53s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 11s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  24m 11s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javac  |  24m 11s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 53s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  21m 53s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   4m 27s |  |  root: The patch generated 0 new + 245 unchanged - 6 fixed = 245 total (was 251)  |
| +1 :green_heart: |  mvnsite  |   5m 19s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   4m  7s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  javadoc  |   4m 37s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   9m 10s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m  8s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 20s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  | 382m  1s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |  34m 58s |  |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 41s |  |  The patch does not generate ASF License warnings.  |
|  |   | 690m 57s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 080e9ebce89a 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / cc4eaaec28e444d9bc27b6ba732e288019deab23 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/4/testReport/ |
| Max. process+thread count | 2453 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4530/4/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java [hunk 8]
```diff
@@ -495,6 +498,94 @@ public static NameNodeMetrics getNameNodeMetrics() {
...
+    }
+    return null;
+  }
+
+  /**
+   * Try to obtain the actual client's machine according to the current user.
+   * @param ipProxyUsers Users who can override client infos.
+   * @return The actual client's machine.
+   */
+  public static String getClientMachine(final String[] ipProxyUsers) {
```
Comment (ferhui) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:544]: one question here. Where do we use this method? Didn't find it.

Comment (ZanderXu) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:544]: Thanks @ferhui  for your review.

`NameNodeRpcServer#getClientMachine()` will use this method as follows:
```
  /**
   * Get the actual client's machine.
   */
  private String getClientMachine() {
    return NameNode.getClientMachine(this.ipProxyUsers);
  }
```

Comment (ferhui) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:544]: Got it, thanks! I didn't load diff for NameNodeRpcServer.java

**Review by ferhui — APPROVED** [review]
It looks good! Thanks!

**Comment by ferhui** [comment]
@ZanderXu Thanks for your great contribution. @goiri Thanks for you review! Merged

**Comment by ZanderXu** [comment]
@ferhui @goiri Thanks for your reivew!

**Pull request merged**